KanesMethod solves kinematics explicitly even if not asked

While dynamic equations are left in explicit form `M [u].diff(t) = f(q, u)`, the kinematic equations are solved explicitly [here](https://github.com/sympy/sympy/blob/0e4bab73562d6f90fcdf5fa079731f26a455347c/sympy/physics/mechanics/kane.py#L259):
```
            f_k = k_kqdot.LUsolve(f_k)
            k_ku = k_kqdot.LUsolve(k_ku)
            k_kqdot = eye(len(qdot))
```

The end result is that in `mass_matrix_full` the columns associated with kinematics are just the identity. This works fine most of the time, but can be unnecessarily slow if an explicit form of the equations is not desired, not to mention that it requires simplification of the explicit right hand side.

Would it make sense to avoid this step unless explicitly requested by the initialization function?

Happy to submit a PR if others here suggest how to best control that behavior.

If we remove the solves, the only relevant failing test I'm getting is this one:

```
        # This code tests our output vs. benchmark values. When r=g=m=1, the
        # critical speed (where all eigenvalues of the linearized equations are 0)
        # is 1 / sqrt(3) for the upright case.
        A = KM.linearize(A_and_B=True)[0]
        A_upright = A.subs({r: 1, g: 1, m: 1}).subs({q1: 0, q2: 0, q3: 0, u1: 0, u3: 0})
        import sympy
>       assert sympy.sympify(A_upright.subs({u2: 1 / sqrt(3)})).eigenvals() == {S.Zero: 6}
E       AssertionError: assert {-sqrt(5)/4 -... 5/8): 1, ...} == {0: 6}
E         Differing items:
E         {0: 2} != {0: 6}
E         Left contains 4 more items:
E         {-1/4 + sqrt(5)/4 - I*sqrt(sqrt(5)/8 + 5/8): 1,
E          -sqrt(5)/4 - 1/4 - I*sqrt(5/8 - sqrt(5)/8): 1,
E          -sqrt(5)/4 - 1/4 + I*sqrt(5/8 - sqrt(5)/8): 1,
E          -1/4 + sqrt(5)/4 + I*sqrt(sqrt(5)/8 + 5/8): 1}...
E         
E         ...Full output truncated (2 lines hidden), use '-vv' to show

sympy/physics/mechanics/tests/test_kane.py:175: AssertionError
```

A_upright ends up with nans and zoos.

--- a/sympy/physics/mechanics/kane.py
+++ b/sympy/physics/mechanics/kane.py
@@ -35,25 +35,32 @@ class KanesMethod(_Methods):
     loads : iterable
         Iterable of (Point, vector) or (ReferenceFrame, vector) tuples
         describing the forces on the system.
-    auxiliary : Matrix
+    auxiliary_eqs : Matrix
         If applicable, the set of auxiliary Kane's
         equations used to solve for non-contributing
         forces.
     mass_matrix : Matrix
-        The system's mass matrix
+        The system's dynamics mass matrix: [k_d; k_dnh]
     forcing : Matrix
-        The system's forcing vector
+        The system's dynamics forcing vector: -[f_d; f_dnh]
     mass_matrix_kin : Matrix
-        The "mass matrix" for kinematic differential equations
+        The "mass matrix" for kinematic differential equations: k_kqdot
     forcing_kin : Matrix
-        The forcing vector for kinematic differential equations
+        The forcing vector for kinematic differential equations: -(k_ku*u + f_k)
     mass_matrix_full : Matrix
-        The "mass matrix" for the u's and q's with explicit kinematics
+        The "mass matrix" for the u's and q's with dynamics and explicit kinematics
     forcing_full : Matrix
-        The "forcing vector" for the u's and q's with explicit kinematics
+        The "forcing vector" for the u's and q's with dynamics and explicit kinematics
 
-    By default, the kinematic and "full" mass matrix and forcing vector are given
-    with the kinematic equations in explicit form. The method also provides an `_implicit` version for those matrices/vectors where the kinematics are left in implicit form
+    Notes
+    =====
+
+    The mass matrices and forcing vectors which involve kinematic equations are given in explicit form.
+    In other words, mass_matrix_kin = k_kqdot = identity
+
+    The Kane object also provides an ``_implicit`` version for those matrices/vectors where the kinematics
+    are left in implicit form, e.g. with mass_matrix_kin_implicit is not necessarily an identity
+    This can provide more compact equations for non-simple kinematics (see https://github.com/sympy/sympy/issues/22626)
 
     Examples
     ========
@@ -559,7 +566,7 @@ class KanesMethod(_Methods):
         constraints) the length of the returned vectors will be o - m + s in
         length. The first o - m equations will be the constrained Kane's
         equations, then the s auxiliary Kane's equations. These auxiliary
-        equations can be accessed with the auxiliary_eqs().
+        equations can be accessed with the auxiliary_eqs property.
 
         Parameters
         ==========
@@ -672,19 +679,22 @@ class KanesMethod(_Methods):
             raise ValueError('Need to compute Fr, Fr* first.')
         return -Matrix([self._f_d, self._f_dnh])
 
+
     def _mass_matrix_kin(self, explicit=True):
         """
-        Returns the "mass matrix" Y_q for the kinematic differential equations
-        Y_q*q' = f_q = Y_u*u + f(q, t)
-        If explicit=True, Y_q = identity
+        Returns the "mass matrix" k_qqdot for the kinematic differential equations
+
+        k_qqdot * q' = f_q = -k_ku * u - f_k(q, t)
+        If explicit=True (default behavior), k_qqdot = identity
         """
         return self._k_kqdot if explicit else self._k_kqdot_implicit
 
     def _forcing_kin(self, explicit=True):
         """
         Returns the "forcing vector" for the kinematic differential equation
-        Y_q*q' = f_q = Y_u*u + f(q, t)
-        If explicit=True (default behavior), assumes Y_q = identity
+
+        k_qqdot * q' = f_q = -k_ku * u - f_k(q, t)
+        If explicit=True (default behavior), implies k_qqdot = identity
         """
         if explicit:
             return -(self._k_ku * Matrix(self.u) + self._f_k)
@@ -693,22 +703,22 @@ class KanesMethod(_Methods):
 
     @property
     def mass_matrix_kin(self):
-        """The mass matrix of the the kinematic differential equations in explicit form."""
+        """The mass matrix of the the kinematic differential equations k_qqdot (explicit form, i.e. k_qqdot=identity)."""
         return self._mass_matrix_kin(True)
 
     @property
     def mass_matrix_kin_implicit(self):
-        """The mass matrix of the the kinematic differential equations in implicit form."""
+        """The mass matrix of the the kinematic differential equations k_qqdot (implicit form)."""
         return self._mass_matrix_kin(False)
 
     @property
     def forcing_kin(self):
-        """The focing vector  the the kinematic differential equations in explicit form."""
+        """The focing vector  the the kinematic differential equations -k_ku * u - f_k (explicit form)."""
         return self._forcing_kin(True)
 
     @property
     def forcing_kin_implicit(self):
-        """The focing vector  the the kinematic differential equations in implicit form."""
+        """The focing vector  the the kinematic differential equations -k_ku * u - f_k (implicit form)."""
         return self._forcing_kin(False)
 
     def _mass_matrix_full(self, explicit=True):

--- a/sympy/physics/mechanics/kane.py
+++ b/sympy/physics/mechanics/kane.py
@@ -43,10 +43,17 @@ class KanesMethod(_Methods):
         The system's mass matrix
     forcing : Matrix
         The system's forcing vector
+    mass_matrix_kin : Matrix
+        The "mass matrix" for kinematic differential equations
+    forcing_kin : Matrix
+        The forcing vector for kinematic differential equations
     mass_matrix_full : Matrix
-        The "mass matrix" for the u's and q's
+        The "mass matrix" for the u's and q's with explicit kinematics
     forcing_full : Matrix
-        The "forcing vector" for the u's and q's
+        The "forcing vector" for the u's and q's with explicit kinematics
+
+    By default, the kinematic and "full" mass matrix and forcing vector are given
+    with the kinematic equations in explicit form. The method also provides an `_implicit` version for those matrices/vectors where the kinematics are left in implicit form
 
     Examples
     ========
@@ -281,6 +288,10 @@ class KanesMethod(_Methods):
                        'coordinates.')
                 raise ValueError(msg.format(nonlin_vars))
 
+            self._f_k_implicit = f_k.xreplace(uaux_zero)
+            self._k_ku_implicit = k_ku.xreplace(uaux_zero)
+            self._k_kqdot_implicit = k_kqdot
+
             # Solve for q'(t) such that the coefficient matrices are now in
             # this form:
             #
@@ -289,19 +300,19 @@ class KanesMethod(_Methods):
             # NOTE : Solving the kinematic differential equations here is not
             # necessary and prevents the equations from being provided in fully
             # implicit form.
-            f_k = k_kqdot.LUsolve(f_k)
-            k_ku = k_kqdot.LUsolve(k_ku)
-            k_kqdot = eye(len(qdot))
-            self._qdot_u_map = dict(zip(qdot, -(k_ku*u + f_k)))
+            f_k_explicit = k_kqdot.LUsolve(f_k)
+            k_ku_explicit = k_kqdot.LUsolve(k_ku)
+            self._qdot_u_map = dict(zip(qdot, -(k_ku_explicit*u + f_k_explicit)))
 
-            self._f_k = f_k.xreplace(uaux_zero)
-            self._k_ku = k_ku.xreplace(uaux_zero)
-            self._k_kqdot = k_kqdot
+            self._f_k = f_k_explicit.xreplace(uaux_zero)
+            self._k_ku = k_ku_explicit.xreplace(uaux_zero)
+            self._k_kqdot = eye(len(qdot))
+
         else:
             self._qdot_u_map = None
-            self._f_k = Matrix()
-            self._k_ku = Matrix()
-            self._k_kqdot = Matrix()
+            self._f_k_implicit = self._f_k = Matrix()
+            self._k_ku_implicit = self._k_ku = Matrix()
+            self._k_kqdot_implicit = self._k_kqdot = Matrix()
 
     def _form_fr(self, fl):
         """Form the generalized active force."""
@@ -581,7 +592,10 @@ class KanesMethod(_Methods):
                 km = KanesMethod(self._inertial, self.q, self._uaux,
                         u_auxiliary=self._uaux, u_dependent=self._udep,
                         velocity_constraints=(self._k_nh * self.u +
-                        self._f_nh))
+                        self._f_nh),
+                        acceleration_constraints=(self._k_dnh * self._udot +
+                        self._f_dnh)
+                        )
             km._qdot_u_map = self._qdot_u_map
             self._km = km
             fraux = km._form_fr(loads)
@@ -652,31 +666,92 @@ class KanesMethod(_Methods):
         return Matrix([self._k_d, self._k_dnh])
 
     @property
-    def mass_matrix_full(self):
-        """The mass matrix of the system, augmented by the kinematic
-        differential equations."""
-        if not self._fr or not self._frstar:
-            raise ValueError('Need to compute Fr, Fr* first.')
-        o = len(self.u)
-        n = len(self.q)
-        return ((self._k_kqdot).row_join(zeros(n, o))).col_join((zeros(o,
-                n)).row_join(self.mass_matrix))
-
-    @property
     def forcing(self):
         """The forcing vector of the system."""
         if not self._fr or not self._frstar:
             raise ValueError('Need to compute Fr, Fr* first.')
         return -Matrix([self._f_d, self._f_dnh])
 
+    def _mass_matrix_kin(self, explicit=True):
+        """
+        Returns the "mass matrix" Y_q for the kinematic differential equations
+        Y_q*q' = f_q = Y_u*u + f(q, t)
+        If explicit=True, Y_q = identity
+        """
+        return self._k_kqdot if explicit else self._k_kqdot_implicit
+
+    def _forcing_kin(self, explicit=True):
+        """
+        Returns the "forcing vector" for the kinematic differential equation
+        Y_q*q' = f_q = Y_u*u + f(q, t)
+        If explicit=True (default behavior), assumes Y_q = identity
+        """
+        if explicit:
+            return -(self._k_ku * Matrix(self.u) + self._f_k)
+        else:
+            return -(self._k_ku_implicit * Matrix(self.u) + self._f_k_implicit)
+
+    @property
+    def mass_matrix_kin(self):
+        """The mass matrix of the the kinematic differential equations in explicit form."""
+        return self._mass_matrix_kin(True)
+
+    @property
+    def mass_matrix_kin_implicit(self):
+        """The mass matrix of the the kinematic differential equations in implicit form."""
+        return self._mass_matrix_kin(False)
+
+    @property
+    def forcing_kin(self):
+        """The focing vector  the the kinematic differential equations in explicit form."""
+        return self._forcing_kin(True)
+
+    @property
+    def forcing_kin_implicit(self):
+        """The focing vector  the the kinematic differential equations in implicit form."""
+        return self._forcing_kin(False)
+
+    def _mass_matrix_full(self, explicit=True):
+        """The mass matrix of the system, augmented by the kinematic
+        differential equations in explicit or implicit form."""
+        if not self._fr or not self._frstar:
+            raise ValueError('Need to compute Fr, Fr* first.')
+        o = len(self.u)
+        n = len(self.q)
+        return (
+            self._mass_matrix_kin(explicit).row_join(zeros(n, o))
+                ).col_join(
+            zeros(o,n).row_join(self.mass_matrix)
+                )
+
+    def _forcing_full(self, explicit=True):
+        """The forcing vector of the system, augmented by the kinematic
+        differential equations in explicit or implicit form."""
+        return Matrix([self._forcing_kin(explicit), self.forcing])
+
+    @property
+    def mass_matrix_full(self):
+        """The mass matrix of the system, augmented by the kinematic
+        differential equations in explicit form."""
+        return self._mass_matrix_full(True)
+
+    @property
+    def mass_matrix_full_implicit(self):
+        """The mass matrix of the system, augmented by the kinematic
+        differential equations in implicit form."""
+        return self._mass_matrix_full(False)
+
     @property
     def forcing_full(self):
         """The forcing vector of the system, augmented by the kinematic
-        differential equations."""
-        if not self._fr or not self._frstar:
-            raise ValueError('Need to compute Fr, Fr* first.')
-        f1 = self._k_ku * Matrix(self.u) + self._f_k
-        return -Matrix([f1, self._f_d, self._f_dnh])
+        differential equations in explicit form."""
+        return self._forcing_full(True)
+
+    @property
+    def forcing_full_implicit(self):
+        """The forcing vector of the system, augmented by the kinematic
+        differential equations in implicit form."""
+        return self._forcing_full(False)
 
     @property
     def q(self):

--- a/sympy/physics/mechanics/kane.py
+++ b/sympy/physics/mechanics/kane.py
@@ -14,7 +14,7 @@ __all__ = ['KanesMethod']
 
 
 class KanesMethod(_Methods):
-    """Kane's method object.
+    r"""Kane's method object.
 
     Explanation
     ===========
@@ -674,7 +674,7 @@ class KanesMethod(_Methods):
 
     @property
     def mass_matrix_kin(self):
-        """The kinematic "mass matrix" $\mathbf{k_{k\dot{q}}}$ of the system."""
+        r"""The kinematic "mass matrix" $\mathbf{k_{k\dot{q}}}$ of the system."""
         return self._k_kqdot.copy() if self.explicit_kinematics else self._k_kqdot_implicit.copy()
 
     @property

--- a/sympy/physics/mechanics/kane.py
+++ b/sympy/physics/mechanics/kane.py
@@ -48,19 +48,24 @@ class KanesMethod(_Methods):
     forcing_kin : Matrix
         The forcing vector for kinematic differential equations: -(k_ku*u + f_k)
     mass_matrix_full : Matrix
-        The "mass matrix" for the u's and q's with dynamics and explicit kinematics
+        The "mass matrix" for the u's and q's with dynamics and kinematics
     forcing_full : Matrix
-        The "forcing vector" for the u's and q's with dynamics and explicit kinematics
+        The "forcing vector" for the u's and q's with dynamics and kinematics
+    explicit_kinematics : bool
+        Boolean whether the mass matrices and forcing vectors should use the
+        explicit form (default) or implicit form for kinematics.
+        See the notes for more details.
 
     Notes
     =====
 
-    The mass matrices and forcing vectors which involve kinematic equations are given in explicit form.
-    In other words, mass_matrix_kin = k_kqdot = identity
-
-    The Kane object also provides an ``_implicit`` version for those matrices/vectors where the kinematics
-    are left in implicit form, e.g. with mass_matrix_kin_implicit is not necessarily an identity
-    This can provide more compact equations for non-simple kinematics (see https://github.com/sympy/sympy/issues/22626)
+    The mass matrices and forcing vectors related to kinematic equations
+    are given in the explicit form by default. In other words, the kinematic
+    mass matrix is $\mathbf{k_{k\dot{q}}} = \mathbf{I}$.
+    In order to get the implicit form of those matrices/vectors, you can set the
+    ``explicit_kinematics`` attribute to ``False``. So $\mathbf{k_{k\dot{q}}}$ is not
+    necessarily an identity matrix. This can provide more compact equations for
+    non-simple kinematics (see #22626).
 
     Examples
     ========
@@ -131,7 +136,7 @@ class KanesMethod(_Methods):
     def __init__(self, frame, q_ind, u_ind, kd_eqs=None, q_dependent=None,
             configuration_constraints=None, u_dependent=None,
             velocity_constraints=None, acceleration_constraints=None,
-            u_auxiliary=None, bodies=None, forcelist=None):
+            u_auxiliary=None, bodies=None, forcelist=None, explicit_kinematics=True):
 
         """Please read the online documentation. """
         if not q_ind:
@@ -147,6 +152,8 @@ class KanesMethod(_Methods):
 
         self._forcelist = forcelist
         self._bodylist = bodies
+
+        self.explicit_kinematics = explicit_kinematics
 
         self._initialize_vectors(q_ind, q_dependent, u_ind, u_dependent,
                 u_auxiliary)
@@ -666,6 +673,19 @@ class KanesMethod(_Methods):
         return self._aux_eq
 
     @property
+    def mass_matrix_kin(self):
+        """The kinematic "mass matrix" $\mathbf{k_{k\dot{q}}}$ of the system."""
+        return self._k_kqdot.copy() if self.explicit_kinematics else self._k_kqdot_implicit.copy()
+
+    @property
+    def forcing_kin(self):
+        """The kinematic "forcing vector" of the system."""
+        if self.explicit_kinematics:
+            return -(self._k_ku * Matrix(self.u) + self._f_k)
+        else:
+            return -(self._k_ku_implicit * Matrix(self.u) + self._f_k_implicit)
+
+    @property
     def mass_matrix(self):
         """The mass matrix of the system."""
         if not self._fr or not self._frstar:
@@ -679,89 +699,21 @@ class KanesMethod(_Methods):
             raise ValueError('Need to compute Fr, Fr* first.')
         return -Matrix([self._f_d, self._f_dnh])
 
-
-    def _mass_matrix_kin(self, explicit=True):
-        """
-        Returns the "mass matrix" k_qqdot for the kinematic differential equations
-
-        k_qqdot * q' = f_q = -k_ku * u - f_k(q, t)
-        If explicit=True (default behavior), k_qqdot = identity
-        """
-        return self._k_kqdot if explicit else self._k_kqdot_implicit
-
-    def _forcing_kin(self, explicit=True):
-        """
-        Returns the "forcing vector" for the kinematic differential equation
-
-        k_qqdot * q' = f_q = -k_ku * u - f_k(q, t)
-        If explicit=True (default behavior), implies k_qqdot = identity
-        """
-        if explicit:
-            return -(self._k_ku * Matrix(self.u) + self._f_k)
-        else:
-            return -(self._k_ku_implicit * Matrix(self.u) + self._f_k_implicit)
-
     @property
-    def mass_matrix_kin(self):
-        """The mass matrix of the the kinematic differential equations k_qqdot (explicit form, i.e. k_qqdot=identity)."""
-        return self._mass_matrix_kin(True)
-
-    @property
-    def mass_matrix_kin_implicit(self):
-        """The mass matrix of the the kinematic differential equations k_qqdot (implicit form)."""
-        return self._mass_matrix_kin(False)
-
-    @property
-    def forcing_kin(self):
-        """The focing vector  the the kinematic differential equations -k_ku * u - f_k (explicit form)."""
-        return self._forcing_kin(True)
-
-    @property
-    def forcing_kin_implicit(self):
-        """The focing vector  the the kinematic differential equations -k_ku * u - f_k (implicit form)."""
-        return self._forcing_kin(False)
-
-    def _mass_matrix_full(self, explicit=True):
+    def mass_matrix_full(self):
         """The mass matrix of the system, augmented by the kinematic
         differential equations in explicit or implicit form."""
         if not self._fr or not self._frstar:
             raise ValueError('Need to compute Fr, Fr* first.')
-        o = len(self.u)
-        n = len(self.q)
-        return (
-            self._mass_matrix_kin(explicit).row_join(zeros(n, o))
-                ).col_join(
-            zeros(o,n).row_join(self.mass_matrix)
-                )
-
-    def _forcing_full(self, explicit=True):
-        """The forcing vector of the system, augmented by the kinematic
-        differential equations in explicit or implicit form."""
-        return Matrix([self._forcing_kin(explicit), self.forcing])
-
-    @property
-    def mass_matrix_full(self):
-        """The mass matrix of the system, augmented by the kinematic
-        differential equations in explicit form."""
-        return self._mass_matrix_full(True)
-
-    @property
-    def mass_matrix_full_implicit(self):
-        """The mass matrix of the system, augmented by the kinematic
-        differential equations in implicit form."""
-        return self._mass_matrix_full(False)
+        o, n = len(self.u), len(self.q)
+        return (self.mass_matrix_kin.row_join(zeros(n, o))).col_join(
+            zeros(o, n).row_join(self.mass_matrix))
 
     @property
     def forcing_full(self):
         """The forcing vector of the system, augmented by the kinematic
-        differential equations in explicit form."""
-        return self._forcing_full(True)
-
-    @property
-    def forcing_full_implicit(self):
-        """The forcing vector of the system, augmented by the kinematic
-        differential equations in implicit form."""
-        return self._forcing_full(False)
+        differential equations in explicit or implicit form."""
+        return Matrix([self.forcing_kin, self.forcing])
 
     @property
     def q(self):

--- a/sympy/physics/mechanics/kane.py
+++ b/sympy/physics/mechanics/kane.py
@@ -675,7 +675,7 @@ class KanesMethod(_Methods):
     @property
     def mass_matrix_kin(self):
         r"""The kinematic "mass matrix" $\mathbf{k_{k\dot{q}}}$ of the system."""
-        return self._k_kqdot.copy() if self.explicit_kinematics else self._k_kqdot_implicit.copy()
+        return self._k_kqdot if self.explicit_kinematics else self._k_kqdot_implicit
 
     @property
     def forcing_kin(self):

--- a/sympy/physics/mechanics/tests/test_kane.py
+++ b/sympy/physics/mechanics/tests/test_kane.py
@@ -82,23 +82,12 @@ def test_two_dof():
     assert KM.explicit_kinematics
     assert KM.mass_matrix_kin == eye(2)
 
-    # Check that the implicit and explicit matrices are different when involving kinematics
-    mf_keys = ['mass_matrix', 'mass_matrix_kin', 'mass_matrix_full',
-               'forcing', 'forcing_kin', 'forcing_full']
-    KM_matrices = {}
-    for explicit_kinematics in [False, True]:
-        KM.explicit_kinematics = explicit_kinematics
-        KM_matrices['explicit' if explicit_kinematics else 'implicit'] = {k:KM.__getattribute__(k) for k in mf_keys}
+    # Check that for the implicit form the mass matrix is not identity
+    KM.explicit_kinematics = False
+    assert KM.mass_matrix_kin == Matrix([[1/2, 0], [0, 2]])
 
-    for k in mf_keys:
-        if k in ['mass_matrix', 'forcing']:
-            # dynamic matrix/vector are unchanged whether we use explicit or implicit kinematics
-            assert KM_matrices['explicit'][k] == KM_matrices['implicit'][k]
-        else:
-            # but matrix/vector involving kinematics are expected to be different
-            assert KM_matrices['explicit'][k] != KM_matrices['implicit'][k]
-
-    # Check that whether using implicit or explicit kinematics the RHS equations are consisten with the matrix form
+    # Check that whether using implicit or explicit kinematics the RHS
+    # equations are consisten with the matrix form
     for explicit_kinematics in [False, True]:
         KM.explicit_kinematics = explicit_kinematics
         assert simplify(KM.rhs() -

--- a/sympy/physics/mechanics/tests/test_kane.py
+++ b/sympy/physics/mechanics/tests/test_kane.py
@@ -77,6 +77,10 @@ def test_two_dof():
     assert simplify(KM.rhs() -
                     KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(4, 1)
 
+    # Check that the implicit form also gives the same result
+    assert simplify(KM.rhs() -
+                    KM.mass_matrix_full_implicit.LUsolve(KM.forcing_full_implicit)) == zeros(4, 1)
+
     # Make sure an error is raised if nonlinear kinematic differential
     # equations are supplied.
     kd = [q1d - u1**2, sin(q2d) - cos(u2)]
@@ -348,3 +352,107 @@ def test_input_format():
     assert expand(rhs[0]) == expand((-k1 * q1 - c1 * u1 + k2 * q2 + c2 * u2)/m)
     assert expand(rhs[1]) == expand((k1 * q1 + c1 * u1 - 2 * k2 * q2 - 2 *
                                     c2 * u2) / m)
+
+
+def test_implicit_kinematics():
+    # Test that implicit kinematics can handle complicated
+    # equations that explicit form struggles with
+    # See https://github.com/sympy/sympy/issues/22626
+
+    # Inertial frame
+    NED = ReferenceFrame('NED')
+    NED_o = Point('NED_o')
+    NED_o.set_vel(NED, 0)
+
+    # body frame
+    q_att = dynamicsymbols('lambda_0:4', real=True)
+    B = NED.orientnew('B', 'Quaternion', q_att)
+
+    # Generalized coordinates
+    q_pos = dynamicsymbols('B_x:z')
+    B_cm = NED_o.locatenew('B_cm', q_pos[0]*B.x + q_pos[1]*B.y + q_pos[2]*B.z)
+
+    q_ind = q_att[1:] + q_pos
+    q_dep = [q_att[0]]
+
+    kinematic_eqs = []
+
+    # Generalized velocities
+    B_ang_vel = B.ang_vel_in(NED)
+    P, Q, R = dynamicsymbols('P Q R')
+    B.set_ang_vel(NED, P*B.x + Q*B.y + R*B.z)
+
+    B_ang_vel_kd = (B.ang_vel_in(NED) - B_ang_vel).simplify()
+
+    # Equating the two gives us the kinematic equation
+    kinematic_eqs += [
+        B_ang_vel_kd & B.x,
+        B_ang_vel_kd & B.y,
+        B_ang_vel_kd & B.z
+    ]
+
+    B_cm_vel = B_cm.vel(NED)
+    U, V, W = dynamicsymbols('U V W')
+    B_cm.set_vel(NED, U*B.x + V*B.y + W*B.z)
+
+    # Compute the velocity of the point using the two methods
+    B_ref_vel_kd = (B_cm.vel(NED) - B_cm_vel)
+
+    # taking dot product with unit vectors to get kinematic equations
+    # relating body coordinates and velocities
+
+    # Note, there is a choice to dot with NED.xyz here. That makes
+    # the implicit form have some bigger terms but is still fine, the
+    # explicit form still struggles though
+    kinematic_eqs += [
+                      B_ref_vel_kd & B.x,
+                      B_ref_vel_kd & B.y,
+                      B_ref_vel_kd & B.z,
+                     ]
+
+    u_ind = [U, V, W, P, Q, R]
+
+    # constraints
+    q_att_vec = Matrix(q_att)
+    config_cons = [q_att_vec.norm()**2 - 1] #unit norm
+    kinematic_eqs = kinematic_eqs + [(q_att_vec.T * q_att_vec.diff())[0]]
+
+    KM = KanesMethod(NED, q_ind, u_ind,
+      q_dependent= q_dep,
+      kd_eqs = kinematic_eqs,
+      configuration_constraints = config_cons,
+      velocity_constraints= [],
+      u_dependent= [], #no dependent speeds
+      u_auxiliary = [] # No auxiliary speeds
+    )
+
+
+    # mass and inertia dyadic relative to CM
+    M_B = symbols('M_B')
+    J_B = inertia(B, *[S(f'J_B_{ax}')*(1 if ax[0] == ax[1] else -1)
+            for ax in ['xx', 'yy', 'zz', 'xy', 'yz', 'xz']])
+    J_B = J_B.subs({S('J_B_xy'): 0, S('J_B_yz'): 0})
+    RB = RigidBody('RB', B_cm, B, M_B, (J_B, B_cm))
+
+    rigid_bodies = [RB]
+    # Forces
+    force_list = [
+        #gravity pointing down
+        (RB.masscenter, RB.mass*S('g')*NED.z),
+        #generic forces and torques in body frame(inputs)
+        (RB.frame, dynamicsymbols('T_z')*B.z),
+        (RB.masscenter, dynamicsymbols('F_z')*B.z)
+    ]
+
+    KM.kanes_equations(rigid_bodies, force_list)
+
+    # Expecting implicit form to be less than 5% of the flops
+    n_ops_explicit = sum(
+        [x.count_ops() for x in KM.forcing_full] +
+        [x.count_ops() for x in KM.mass_matrix_full]
+    )
+    n_ops_implicit = sum(
+        [x.count_ops() for x in KM.forcing_full_implicit] +
+        [x.count_ops() for x in KM.mass_matrix_full_implicit]
+    )
+    assert n_ops_implicit / n_ops_explicit < .05

--- a/sympy/physics/mechanics/tests/test_kane.py
+++ b/sympy/physics/mechanics/tests/test_kane.py
@@ -5,6 +5,7 @@ from sympy.physics.mechanics import (dynamicsymbols, ReferenceFrame, Point,
                                      RigidBody, KanesMethod, inertia, Particle,
                                      dot)
 from sympy.testing.pytest import raises
+from sympy.core.backend import USE_SYMENGINE
 
 
 def test_one_dof():
@@ -414,18 +415,24 @@ def test_implicit_kinematics():
 
     # constraints
     q_att_vec = Matrix(q_att)
-    config_cons = [q_att_vec.norm()**2 - 1] #unit norm
+    config_cons = [(q_att_vec.T*q_att_vec)[0] - 1] #unit norm
     kinematic_eqs = kinematic_eqs + [(q_att_vec.T * q_att_vec.diff())[0]]
 
-    KM = KanesMethod(NED, q_ind, u_ind,
-      q_dependent= q_dep,
-      kd_eqs = kinematic_eqs,
-      configuration_constraints = config_cons,
-      velocity_constraints= [],
-      u_dependent= [], #no dependent speeds
-      u_auxiliary = [] # No auxiliary speeds
-    )
-
+    try:
+        KM = KanesMethod(NED, q_ind, u_ind,
+          q_dependent= q_dep,
+          kd_eqs = kinematic_eqs,
+          configuration_constraints = config_cons,
+          velocity_constraints= [],
+          u_dependent= [], #no dependent speeds
+          u_auxiliary = [] # No auxiliary speeds
+        )
+    except Exception as e:
+        # symengine struggles with these kinematic equations
+        if USE_SYMENGINE and 'Matrix is rank deficient' in str(e):
+            return
+        else:
+            raise e
 
     # mass and inertia dyadic relative to CM
     M_B = symbols('M_B')

--- a/sympy/physics/mechanics/tests/test_kane.py
+++ b/sympy/physics/mechanics/tests/test_kane.py
@@ -1,3 +1,4 @@
+from sympy import solve
 from sympy.core.backend import (cos, expand, Matrix, sin, symbols, tan, sqrt, S,
                                 zeros)
 from sympy.simplify.simplify import simplify
@@ -463,3 +464,37 @@ def test_implicit_kinematics():
         [x.count_ops() for x in KM.mass_matrix_full_implicit]
     )
     assert n_ops_implicit / n_ops_explicit < .05
+
+    # Ideally we would check that implicit and explicit equations give the same result as done in test_one_dof
+    # But the whole raison-d'etre of the implicit equations is to deal with problems such
+    # as this one where the explicit form is too complicated to handle, especially the angular part
+    # (i.e. tests would be too slow)
+    # Instead, we check that the kinematic equations are correct using more fundamental tests:
+    #
+    # (1) that we recover the kinematic equations we have provided
+    assert (KM.mass_matrix_kin_implicit * KM.q.diff() - KM.forcing_kin_implicit) == Matrix(kinematic_eqs)
+
+    # (2) that rate of quaternions matches what 'textbook' solutions give
+    # Note that we just use the explicit kinematics for the linear velocities
+    # as they are not as complicated as the angular ones
+    qdot_candidate = KM.forcing_kin
+
+    quat_dot_textbook = Matrix([
+        [0, -P, -Q, -R],
+        [P,  0,  R, -Q],
+        [Q, -R,  0,  P],
+        [R,  Q, -P,  0],
+    ]) * q_att_vec / 2
+
+    # Again, if we don't use this "textbook" solution
+    # sympy will struggle to deal with the terms related to quaternion rates
+    # due to the number of operations involved
+    qdot_candidate[-1] = quat_dot_textbook[0] # lambda_0, note the [-1] as sympy's Kane puts the dependent coordinate last
+    qdot_candidate[0]  = quat_dot_textbook[1] # lambda_1
+    qdot_candidate[1]  = quat_dot_textbook[2] # lambda_2
+    qdot_candidate[2]  = quat_dot_textbook[3] # lambda_3
+
+    # sub the config constraint in the candidate solution and compare to the implicit rhs
+    lambda_0_sol = solve(config_cons[0], q_att_vec[0])[1]
+    lhs_candidate = simplify(KM.mass_matrix_kin_implicit * qdot_candidate).subs({q_att_vec[0]: lambda_0_sol})
+    assert lhs_candidate == KM.forcing_kin_implicit

--- a/sympy/physics/mechanics/tests/test_kane.py
+++ b/sympy/physics/mechanics/tests/test_kane.py
@@ -84,7 +84,7 @@ def test_two_dof():
 
     # Check that for the implicit form the mass matrix is not identity
     KM.explicit_kinematics = False
-    assert KM.mass_matrix_kin == Matrix([[1/2, 0], [0, 2]])
+    assert KM.mass_matrix_kin == Matrix([[S(1)/2, 0], [0, 2]])
 
     # Check that whether using implicit or explicit kinematics the RHS
     # equations are consisten with the matrix form

--- a/sympy/physics/mechanics/tests/test_kane.py
+++ b/sympy/physics/mechanics/tests/test_kane.py
@@ -1,6 +1,6 @@
 from sympy import solve
 from sympy.core.backend import (cos, expand, Matrix, sin, symbols, tan, sqrt, S,
-                                zeros)
+                                zeros, eye)
 from sympy.simplify.simplify import simplify
 from sympy.physics.mechanics import (dynamicsymbols, ReferenceFrame, Point,
                                      RigidBody, KanesMethod, inertia, Particle,
@@ -54,7 +54,9 @@ def test_two_dof():
     P2 = Point('P2')
     P1.set_vel(N, u1 * N.x)
     P2.set_vel(N, (u1 + u2) * N.x)
-    kd = [q1d - u1, q2d - u2]
+    # Note we multiply the kinematic equation by an arbitrary factor
+    # to test the implicit vs explicit kinematics attribute
+    kd = [q1d/2 - u1/2, 2*q2d - 2*u2]
 
     # Now we create the list of forces, then assign properties to each
     # particle, then create a list of all particles.
@@ -76,12 +78,31 @@ def test_two_dof():
     assert expand(rhs[1]) == expand((k1 * q1 + c1 * u1 - 2 * k2 * q2 - 2 *
                                     c2 * u2) / m)
 
-    assert simplify(KM.rhs() -
-                    KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(4, 1)
+    # Check that the explicit form is the default and kinematic mass matrix is identity
+    assert KM.explicit_kinematics
+    assert KM.mass_matrix_kin == eye(2)
 
-    # Check that the implicit form also gives the same result
-    assert simplify(KM.rhs() -
-                    KM.mass_matrix_full_implicit.LUsolve(KM.forcing_full_implicit)) == zeros(4, 1)
+    # Check that the implicit and explicit matrices are different when involving kinematics
+    mf_keys = ['mass_matrix', 'mass_matrix_kin', 'mass_matrix_full',
+               'forcing', 'forcing_kin', 'forcing_full']
+    KM_matrices = {}
+    for explicit_kinematics in [False, True]:
+        KM.explicit_kinematics = explicit_kinematics
+        KM_matrices['explicit' if explicit_kinematics else 'implicit'] = {k:KM.__getattribute__(k) for k in mf_keys}
+
+    for k in mf_keys:
+        if k in ['mass_matrix', 'forcing']:
+            # dynamic matrix/vector are unchanged whether we use explicit or implicit kinematics
+            assert KM_matrices['explicit'][k] == KM_matrices['implicit'][k]
+        else:
+            # but matrix/vector involving kinematics are expected to be different
+            assert KM_matrices['explicit'][k] != KM_matrices['implicit'][k]
+
+    # Check that whether using implicit or explicit kinematics the RHS equations are consisten with the matrix form
+    for explicit_kinematics in [False, True]:
+        KM.explicit_kinematics = explicit_kinematics
+        assert simplify(KM.rhs() -
+                        KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(4, 1)
 
     # Make sure an error is raised if nonlinear kinematic differential
     # equations are supplied.
@@ -426,7 +447,8 @@ def test_implicit_kinematics():
           configuration_constraints = config_cons,
           velocity_constraints= [],
           u_dependent= [], #no dependent speeds
-          u_auxiliary = [] # No auxiliary speeds
+          u_auxiliary = [], # No auxiliary speeds
+          explicit_kinematics = False # implicit kinematics
         )
     except Exception as e:
         # symengine struggles with these kinematic equations
@@ -455,14 +477,21 @@ def test_implicit_kinematics():
     KM.kanes_equations(rigid_bodies, force_list)
 
     # Expecting implicit form to be less than 5% of the flops
+    n_ops_implicit = sum(
+        [x.count_ops() for x in KM.forcing_full] +
+        [x.count_ops() for x in KM.mass_matrix_full]
+    )
+    # Save implicit kinematic matrices to use later
+    mass_matrix_kin_implicit = KM.mass_matrix_kin
+    forcing_kin_implicit = KM.forcing_kin
+
+    KM.explicit_kinematics = True
     n_ops_explicit = sum(
         [x.count_ops() for x in KM.forcing_full] +
         [x.count_ops() for x in KM.mass_matrix_full]
     )
-    n_ops_implicit = sum(
-        [x.count_ops() for x in KM.forcing_full_implicit] +
-        [x.count_ops() for x in KM.mass_matrix_full_implicit]
-    )
+    forcing_kin_explicit = KM.forcing_kin
+
     assert n_ops_implicit / n_ops_explicit < .05
 
     # Ideally we would check that implicit and explicit equations give the same result as done in test_one_dof
@@ -472,12 +501,12 @@ def test_implicit_kinematics():
     # Instead, we check that the kinematic equations are correct using more fundamental tests:
     #
     # (1) that we recover the kinematic equations we have provided
-    assert (KM.mass_matrix_kin_implicit * KM.q.diff() - KM.forcing_kin_implicit) == Matrix(kinematic_eqs)
+    assert (mass_matrix_kin_implicit * KM.q.diff() - forcing_kin_implicit) == Matrix(kinematic_eqs)
 
     # (2) that rate of quaternions matches what 'textbook' solutions give
     # Note that we just use the explicit kinematics for the linear velocities
     # as they are not as complicated as the angular ones
-    qdot_candidate = KM.forcing_kin
+    qdot_candidate = forcing_kin_explicit
 
     quat_dot_textbook = Matrix([
         [0, -P, -Q, -R],
@@ -496,5 +525,5 @@ def test_implicit_kinematics():
 
     # sub the config constraint in the candidate solution and compare to the implicit rhs
     lambda_0_sol = solve(config_cons[0], q_att_vec[0])[1]
-    lhs_candidate = simplify(KM.mass_matrix_kin_implicit * qdot_candidate).subs({q_att_vec[0]: lambda_0_sol})
-    assert lhs_candidate == KM.forcing_kin_implicit
+    lhs_candidate = simplify(mass_matrix_kin_implicit * qdot_candidate).subs({q_att_vec[0]: lambda_0_sol})
+    assert lhs_candidate == forcing_kin_implicit

